In [1]:
import numpy as np 
import urdf2casadi.urdfparser as u2c
from urdf2casadi.geometry import plucker
from urdf_parser_py.urdf import URDF, Pose
import PyKDL as kdl
import kdl_parser_py.urdf as kdlurdf
from timeit import Timer, timeit, repeat
import rbdl
import pybullet as pb


In [2]:
def median(lst):
    n = len(lst)
    if n < 1:
            return None
    if n % 2 == 1:
            return sorted(lst)[n//2]
    else:
            return sum(sorted(lst)[n//2-1:n//2+1])/2.0
        
def average(lst): 
    return sum(lst) / len(lst) 


def c_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToCoriolis(q_kdl, qdot_kdl, g_kdl)


def c_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    c_u2c(q_none, qdot_none)
    
def c_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    c_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot_none, qddot)
    
    
def c_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.NonlinearEffects(rbdlmodel, q_np, qdot_np, c_rbdl)


def g_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToGravity(q_kdl, g_kdl)


def g_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_u2c(q_none)
    
def g_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot, qddot)
    
def g_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.InverseDynamics(rbdlmodel, q_np, qdot_np, qddot_np, g_rbdl)
    
    
def id_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    id_u2c(q_none, qdot_none, qddot_none)
    
    
def id_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2 
    id_pb = pb.calculateInverseDynamics(pbmodel, q_none, qdot_none, qddot_none)
    
    
def id_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.InverseDynamics(rbdlmodel, q_np, qdot_np, qddot_np, id_rbdl)

    
def m_kdl_func():
    for j in range(njoints):
        q_kdl[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    kdl.ChainDynParam(kdlmodel, gravity_kdl).JntToMass(q_kdl, m_kdl)


def m_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    m_u2c(q_none)
    
def m_pb_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    m = pb.calculateMassMatrix(pbmodel, q_none)
    
    
def m_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.CompositeRigidBodyAlgorithm(rbdlmodel, q_np, m_rbdl)
    
    
def aba_u2c_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_aba_u2c(q_none, qdot_none, tau_none)
    
    
def crba_u2c_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_crba_u2c(q_none, qdot_none, tau_none)
    
    
def fd_rbdl_func():
    for j in range(njoints):
        q_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_np[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    rbdl.ForwardDynamics(rbdlmodel, q_np, qdot_np, tau_np, qddot_np)
    

In [4]:
#pbmodel = pb.loadURDF("pantilt.urdf")
sim = pb.connect(pb.DIRECT)
#pbmodel = pb.loadURDF("/urdf4timing/1dof.urdf")
ndofs = 10
urdf_nr = list(range(ndofs + 1))
urdf_nr.pop(0)


nitr = 1000
nrepeat = 100
    

path_to_urdf = "/home/lmjohann/urdf2casadi/examples/urdf/pantilt.urdf"
root = 'base_link'
tip = "tilt_link"

ok, kdl_tree = kdlurdf.treeFromFile(path_to_urdf)
kdlmodel = kdl_tree.getChain(root,tip)

rbdlmodel = rbdl.loadModel(path_to_urdf)
pbmodel = pb.loadURDF(path_to_urdf, useFixedBase=True, flags = pb.URDF_USE_INERTIA_FROM_FILE)
asd = u2c.URDFparser()
robot_desc = asd.from_file(path_to_urdf)

jlist, names, qmax, qmin = asd.get_joint_info(root, tip)
njoints = asd.get_n_joints(root, tip)

gravity_kdl = kdl.Vector()
gravity_kdl[2] = -9.81
gravity_u2c = [0, 0, -9.81]

#rbdl declarations
q_np = np.zeros(njoints)
qdot_np = np.zeros(njoints)
qddot_np = np.zeros(njoints)
g_rbdl = np.zeros(njoints)
c_rbdl = np.zeros(njoints)
id_rbdl = np.zeros(njoints)

#kdl declatations
q_kdl = kdl.JntArray(njoints)
qdot_kdl = kdl.JntArray(njoints)
g_kdl = kdl.JntArray(njoints)

#u2c and pybullet declarations
q_none = [None]*njoints
qdot_none = [None]*njoints
qddot_none = [None]*njoints
g_pb = [None]*njoints

qddot = []
for i in range(njoints):
    qddot.append(0.)

qdot = []
for i in range(njoints):
    qdot.append(0.)
    
    
m_kdl = kdl.JntSpaceInertiaMatrix(njoints)
m_rbdl = (rbdlmodel.q_size, rbdlmodel.q_size)
m_rbdl = np.zeros(m_rbdl)
tau_np = np.zeros(njoints)
tau_none = [None]*njoints

g_u2c = asd.get_gravity_rnea(root, tip, gravity_u2c)
c_u2c = asd.get_coriolis_rnea(root, tip)
m_u2c = asd.get_inertia_matrix_crba(root, tip)
fd_aba_u2c = asd.get_forward_dynamics_aba(root, tip, gravity_u2c)
fd_crba_u2c = asd.get_forward_dynamics_crba(root, tip, gravity_u2c)
id_u2c = asd.get_inverse_dynamics_rnea(root, tip, gravity_u2c)




The root link base_link has an inertia specified in the URDF, but KDL does not support a root link with an inertia.  As a workaround, you can add an extra dummy link to your URDF.
/home/lmjohann/urdf2casadi/examples/urdf/pantilt.urdf


In [4]:
timeit_g_u2c = repeat("g_u2c_func()", setup = "from __main__ import g_u2c_func", repeat = nrepeat, number = nitr)
mediantime_g_u2c  = median(timeit_g_u2c)/nitr*1000000

timeit_g_pb = repeat("g_pb_func()", setup = "from __main__ import g_pb_func", repeat = nrepeat, number = nitr)
mediantime_g_pb = median(timeit_g_pb)/nitr*1000000

timeit_g_rbdl = repeat("g_rbdl_func()", setup = "from __main__ import g_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_g_rbdl = median(timeit_g_rbdl)/nitr*1000000

timeit_g_kdl = repeat("g_kdl_func()", setup = "from __main__ import g_kdl_func", repeat = nrepeat, number = nitr)
mediantime_g_kdl = median(timeit_g_kdl)/nitr*1000000


timeit_c_u2c = repeat("c_u2c_func()", setup = "from __main__ import c_u2c_func", repeat = nrepeat, number = nitr)
mediantime_c_u2c  = median(timeit_c_u2c)/nitr*1000000

timeit_c_pb = repeat("c_pb_func()", setup = "from __main__ import c_pb_func", repeat = nrepeat, number = nitr)
mediantime_c_pb = median(timeit_c_pb)/nitr*1000000

timeit_c_rbdl = repeat("c_rbdl_func()", setup = "from __main__ import c_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_c_rbdl = median(timeit_c_rbdl)/nitr*1000000

timeit_c_kdl = repeat("c_kdl_func()", setup = "from __main__ import c_kdl_func", repeat = nrepeat, number = nitr)
mediantime_c_kdl = median(timeit_c_kdl)/nitr*1000000

timeit_id_u2c = repeat("id_u2c_func()", setup = "from __main__ import id_u2c_func", repeat = nrepeat, number = nitr)
mediantime_id_u2c  = median(timeit_id_u2c)/nitr*1000000

timeit_id_pb = repeat("id_pb_func()", setup = "from __main__ import id_pb_func", repeat = nrepeat, number = nitr)
mediantime_id_pb = median(timeit_id_pb)/nitr*1000000

timeit_id_rbdl = repeat("id_rbdl_func()", setup = "from __main__ import id_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_id_rbdl = median(timeit_id_rbdl)/nitr*1000000


timeit_m_u2c = repeat("m_u2c_func()", setup = "from __main__ import m_u2c_func", repeat = nrepeat, number = nitr)
mediantime_m_u2c  = median(timeit_m_u2c)/nitr*1000000

timeit_m_pb = repeat("m_pb_func()", setup = "from __main__ import m_pb_func", repeat = nrepeat, number = nitr)
mediantime_m_pb = median(timeit_m_pb)/nitr*1000000

timeit_m_rbdl = repeat("m_rbdl_func()", setup = "from __main__ import m_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_m_rbdl = median(timeit_m_rbdl)/nitr*1000000

timeit_m_kdl = repeat("m_kdl_func()", setup = "from __main__ import m_kdl_func", repeat = nrepeat, number = nitr)
mediantime_m_kdl = median(timeit_m_kdl)/nitr*1000000


timeit_fdaba_u2c = repeat("aba_u2c_func()", setup = "from __main__ import aba_u2c_func", repeat = nrepeat, number = nitr)
mediantime_fdaba_u2c = median(timeit_fdaba_u2c)/nitr*1000000

timeit_fdcrba_u2c = repeat("crba_u2c_func()", setup = "from __main__ import crba_u2c_func", repeat = nrepeat, number = nitr)
mediantime_fdcrba_u2c = median(timeit_fdcrba_u2c)/nitr*1000000

timeit_fd_rbdl = repeat("fd_rbdl_func()", setup = "from __main__ import fd_rbdl_func", repeat = nrepeat, number = nitr)
mediantime_fd_rbdl = median(timeit_fd_rbdl)/nitr*1000000


In [5]:
print "\npendulum_g_u2c=", mediantime_g_u2c
print "\npendulum_id_u2c=", mediantime_id_u2c
print "\npendulum_c_u2c=", mediantime_c_u2c
print "\npendulum_m_u2c=", mediantime_m_u2c

print "\npendulum_g_rbdl=", mediantime_g_rbdl
print "\npendulum_id_rbdl=", mediantime_id_rbdl
print "\npendulum_c_rbdl=", mediantime_c_rbdl
print "\npendulum_m_rbdl=", mediantime_m_rbdl

print "\npendulum_g_pb=", mediantime_g_pb
print "\npendulum_id_pb=", mediantime_id_pb
print "\npendulum_c_pb=", mediantime_c_pb
print "\npendulum_m_pb=", mediantime_m_pb

print "\npendulum_g_kdl=", mediantime_g_kdl
print "\npendulum_c_kdl=", mediantime_c_kdl
print "\npendulum_m_kdl=", mediantime_m_kdl

print "\nfdrbdl=", mediantime_fd_rbdl
print "\nfdcrba=", mediantime_fdcrba_u2c 
print "\nfdaba=", mediantime_fdaba_u2c


pendulum_g_u2c= 13.0249261856

pendulum_id_u2c= 28.7500619888

pendulum_c_u2c= 20.6990242004

pendulum_m_u2c= 12.4354362488

pendulum_g_rbdl= 3.01766395569

pendulum_id_rbdl= 4.91309165955

pendulum_c_rbdl= 3.62646579742

pendulum_m_rbdl= 2.20990180969

pendulum_g_pb= 2.30598449707

pendulum_id_pb= 4.00280952454

pendulum_c_pb= 3.19814682007

pendulum_m_pb= 2.23708152771

pendulum_g_kdl= 4.9455165863

pendulum_c_kdl= 6.08956813812

pendulum_m_kdl= 4.90736961365

fdrbdl= 5.14590740204

fdcrba= 28.6569595337

fdaba= 28.6965370178



pendulum_g_u2c= 12.6429796219

pendulum_id_u2c= 28.8081169128

pendulum_C_u2c= 20.6830501556

pendulum_M_u2c= 13.1369829178

fdcrba= 28.9306640625

fdaba= 28.787612915

# Derivatives

## q

In [6]:
import casadi as cs

qs =cs.SX.sym("qs",njoints)
qsdot =cs.SX.sym("qsdot",njoints)
qsddot =cs.SX.sym("qsddot",njoints)
taus =cs.SX.sym("taus",njoints)

c_dq = cs.jacobian(c_u2c(qs, qsdot), qs)
g_dq = cs.jacobian(g_u2c(qs), qs)
m_dq = cs.jacobian(m_u2c(qs), qs)
id_dq = cs.jacobian(id_u2c(qs, qsdot, qsddot), qs)
fd_aba_dq = cs.jacobian(fd_aba_u2c(qs, qsdot, taus), qs)
fd_crba_dq = cs.jacobian(fd_crba_u2c(qs, qsdot, taus), qs)


g_dq = cs.Function("g_dq", [qs], [g_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})
c_dq = cs.Function("C_dq", [qs, qsdot], [c_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})
m_dq = cs.Function("M_dq", [qs], [m_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})
id_dq = cs.Function("id_dq", [qs, qsdot, qsddot], [id_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_dq = cs.Function("fd_aba_dq", [qs, qsdot, taus], [fd_aba_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_dq = cs.Function("fd_crba_dq", [qs, qsdot, taus], [fd_crba_dq], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [7]:
def fdcrba_dq_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_crba_dq(q_none, qdot_none, tau_none)

def fdaba_dq_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    fd_aba_dq(q_none, qdot_none, tau_none)
    
    
def m_dq_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2   
    m_dq(q_none)
    
def g_dq_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    g_dq(q_none)


def c_dq_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    c_dq(q_none, qdot_none)
    
def id_dq_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    id_dq(q_none, qdot_none, qddot_none)

def aba_dq_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    fd_aba_dq(q_none, qdot_none, tau_none)
    
def crba_dq_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    
    fd_crba_dq(q_none, qdot_none, tau_none)

In [8]:
timeit_idj = repeat("id_dq_func()", setup = "from __main__ import id_dq_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000

timeit_gj = repeat("g_dq_func()", setup = "from __main__ import g_dq_func", repeat = nrepeat, number = nitr)
mediantime_gj = median(timeit_gj)/nitr*1000000

timeit_cj = repeat("c_dq_func()", setup = "from __main__ import c_dq_func", repeat = nrepeat, number = nitr)
mediantime_cj = median(timeit_cj)/nitr*1000000

timeit_mj = repeat("m_dq_func()", setup = "from __main__ import m_dq_func", repeat = nrepeat, number = nitr)
mediantime_mj = median(timeit_mj)/nitr*1000000

timeit_fdabaj = repeat("aba_dq_func()", setup = "from __main__ import aba_dq_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("crba_dq_func()", setup = "from __main__ import crba_dq_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\npendulum_id_d=", mediantime_idj
print "\npendulum_g_d=", mediantime_gj
print "\npendulum_c_d=", mediantime_cj
print "\npendulum_m_d=", mediantime_mj
print "\npendulum_fdaba_d=", mediantime_fd_abaj
print "\npendulum_fdcrba_d=", mediantime_fd_crbaj


pendulum_id_d= 28.3290147781

pendulum_g_d= 12.4489068985

pendulum_c_d= 20.2829837799

pendulum_m_d= 12.4139785767

pendulum_fdaba_d= 28.4889936447

pendulum_fdcrba_d= 28.4065008163



pendulum_id_d= 28.7179946899

pendulum_g_d= 12.5154256821

pendulum_C_d= 20.6383466721

pendulum_M_d= 12.5584602356

pendulum_fdaba_d= 28.6189317703

pendulum_fdcrba_d= 28.6304950714


## qdot


In [7]:
import casadi as cs
qs =cs.SX.sym("qs",njoints)
qsdot =cs.SX.sym("qsdot",njoints)
qsddot =cs.SX.sym("qsddot",njoints)
taus =cs.SX.sym("taus",njoints)

c_dqdot = cs.jacobian(c_u2c(qs, qsdot), qs)
id_dqdot = cs.jacobian(id_u2c(qs, qsdot, qsddot), qsdot)
fd_aba_dqdot = cs.jacobian(fd_aba_u2c(qs, qsdot, taus), qsdot)
fd_crba_dqdot = cs.jacobian(fd_crba_u2c(qs, qsdot, taus), qsdot)

c_dqdot = cs.Function("c_dq", [qs, qsdot], [c_dqdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})
id_dqdot = cs.Function("id_dq", [qs, qsdot, qsddot], [id_dqdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_dqdot = cs.Function("fd_aba_dq", [qs, qsdot, taus], [fd_aba_dqdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_dqdot = cs.Function("fd_crba_dq", [qs, qsdot, taus], [fd_crba_dqdot], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [8]:


def c_dqdot_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    c_dqdot(q_none, qdot_none)
    
def id_dqdot_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    id_dqdot(q_none, qdot_none, qddot_none)

def aba_dqdot_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    fd_aba_dqdot(q_none, qdot_none, tau_none)
    
def crba_dqdot_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    
    fd_crba_dqdot(q_none, qdot_none, tau_none)

In [12]:
timeit_idj = repeat("id_dqdot_func()", setup = "from __main__ import id_dqdot_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000

timeit_cj = repeat("c_dqdot_func()", setup = "from __main__ import c_dqdot_func", repeat = nrepeat, number = nitr)
mediantime_cj = median(timeit_cj)/nitr*1000000


timeit_fdabaj = repeat("aba_dqdot_func()", setup = "from __main__ import aba_dqdot_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("crba_dqdot_func()", setup = "from __main__ import crba_dqdot_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\npendulum_id_d=", mediantime_idj
print "\npendulum_C_d=", mediantime_cj
print "\npendulum_fdaba_d=", mediantime_fd_abaj
print "\npendulum_fdcrba_d=", mediantime_fd_crbaj


pendulum_id_d= 29.2010307312

pendulum_C_d= 20.8079814911

pendulum_fdaba_d= 30.0650596619

pendulum_fdcrba_d= 30.0455093384




pendulum_id_d= 28.6995172501

pendulum_C_d= 20.6029415131

pendulum_fdaba_d= 28.5149812698

pendulum_fdcrba_d= 28.4414291382


## qddot / tau


In [12]:


id_dqddot = cs.jacobian(id_u2c(qs, qsdot, qsddot), qsddot)
fd_aba_dtau = cs.jacobian(fd_aba_u2c(qs, qsdot, taus), taus)
fd_crba_dtau = cs.jacobian(fd_crba_u2c(qs, qsdot, taus), taus)

id_dqddot = cs.Function("id_qddot", [qs, qsdot, qsddot], [id_dqddot], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_aba_dtau = cs.Function("fd_aba_dtau", [qs, qsdot, taus], [fd_aba_dtau], {"jit": True, "jit_options":{"flags":"-Ofast"}})
fd_crba_dtau = cs.Function("fd_crba_dtau", [qs, qsdot, taus], [fd_crba_dtau], {"jit": True, "jit_options":{"flags":"-Ofast"}})

In [13]:
def id_dqddot_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    id_dqddot(q_none, qdot_none, qddot_none)

def aba_dtau_func():   
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    fd_aba_dtau(q_none, qdot_none, tau_none)
    
def crba_dtau_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    
    fd_crba_dtau(q_none, qdot_none, tau_none)

In [17]:
timeit_idj = repeat("id_dqddot_func()", setup = "from __main__ import id_dqddot_func", repeat = nrepeat, number = nitr)
mediantime_idj = median(timeit_idj)/nitr*1000000

timeit_fdabaj = repeat("aba_dtau_func()", setup = "from __main__ import aba_dtau_func", repeat = nrepeat, number = nitr)
mediantime_fd_abaj = median(timeit_fdabaj)/nitr*1000000

timeit_fdcrbaj = repeat("crba_dtau_func()", setup = "from __main__ import crba_dtau_func", repeat = nrepeat, number = nitr)
mediantime_fd_crbaj = median(timeit_fdcrbaj)/nitr*1000000

   
print "\npendulum_id_d=", mediantime_idj
print "\npendulum_fdaba_d=", mediantime_fd_abaj
print "\npendulum_fdcrba_d=", mediantime_fd_crbaj


pendulum_id_d= 29.6385288239

pendulum_fdaba_d= 28.5230875015

pendulum_fdcrba_d= 28.6674499512


In [ ]:
pendulum_id_d= 28.235077858

pendulum_fdaba_d= 28.3834934235

pendulum_fdcrba_d= 28.3510684967

## derivatives all 

In [5]:
import casadi as cs

qs =cs.SX.sym("qs",njoints)
qsdot =cs.SX.sym("qsdot",njoints)
qsddot =cs.SX.sym("qsddot",njoints)
qsdddot = cs.SX.sym("qsdddot", njoints)
taus =cs.SX.sym("taus",njoints)
tausdot =cs.SX.sym("tausdot",njoints)

c_vec = cs.vertcat(qs, qsdot)
c_dvec = cs.vertcat(qsdot, qsddot)
id_vec = cs.vertcat(qs, qsdot, qsddot)
id_dvec = cs.vertcat(qsdot, qsdot, qsdddot)
fd_vec = cs.vertcat(qs, qsdot, qsddot)
fd_dvec = cs.vertcat(qsdot, qsdot, tausdot)

dc = cs.jacobian(c_u2c(qs, qsdot), c_vec)
dc = cs.Function("dc", [qs, qsdot], [dc], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dctimes = cs.jtimes(c_u2c(qs, qsdot), c_vec, c_dvec)
dctimes = cs.Function("dctimes", [qs, qsdot, qsddot], [dctimes], {"jit": True, "jit_options":{"flags":"-Ofast"}})

did = cs.jacobian(id_u2c(qs, qsdot, qsddot), id_vec)
did = cs.Function("did", [qs, qsdot, qsddot], [did], {"jit": True, "jit_options":{"flags":"-Ofast"}})

didtimes = cs.jtimes(id_u2c(qs, qsdot, qsddot), id_vec, id_dvec)
didtimes = cs.Function("didtimes", [qs, qsdot, qsddot, qsdddot], [didtimes], {"jit": True, "jit_options":{"flags":"-Ofast"}})


dfd_aba = cs.jacobian(fd_aba_u2c(qs, qsdot, taus), fd_vec)
dfd_aba = cs.Function("dfd_aba", [qs, qsdot, taus], [dfd_aba], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dfdtimes_aba = cs.jtimes(fd_aba_u2c(qs, qsdot, taus), fd_vec, fd_dvec)
dfdtimes_aba = cs.Function("dfdtimes_aba", [qs, qsdot, taus, qsddot, tausdot], [dfdtimes_aba], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dfd_crba = cs.jacobian(fd_crba_u2c(qs, qsdot, taus), fd_vec)
dfd_crba = cs.Function("dfd_crba", [qs, qsdot, taus], [dfd_crba], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dfdtimes_crba = cs.jtimes(fd_crba_u2c(qs, qsdot, taus), fd_vec, fd_dvec)
dfdtimes_crba = cs.Function("dfdtimes_crba", [qs, qsdot, taus, qsddot, tausdot], [dfdtimes_crba], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dg = cs.jacobian(g_u2c(qs), qs)
dg = cs.Function("dg", [qs], [dg], {"jit": True, "jit_options":{"flags":"-Ofast"}})


dm = cs.jacobian(m_u2c(qs), qs)
dm = cs.Function("dm", [qs], [dm], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dgtimes = cs.jtimes(g_u2c(qs), qs, qsdot)
dgtimes = cs.Function("dgtimes", [qs, qsdot], [dgtimes], {"jit": True, "jit_options":{"flags":"-Ofast"}})

dmtimes = cs.jtimes(m_u2c(qs), qs, qsdot)
dmtimes = cs.Function("dmtimes", [qs, qsdot], [dmtimes], {"jit": True, "jit_options":{"flags":"-Ofast"}})


print "\npendulum_nodes_did = ", did.n_nodes()
print "\npendulum_nodes_dc = ", dc.n_nodes()
print "\npendulum_nodes_dm = ", dm.n_nodes()
print "\npendulum_nodes_dg = ", dg.n_nodes()
print "\npendulum_nodes_dfd_aba = ", dfd_aba.n_nodes()
print "\npendulum_nodes_dfd_crba =", dfd_crba.n_nodes()

print "\npendulum_nodes_didtimes = ", didtimes.n_nodes()
print "\npendulum_nodes_dctimes = ", dctimes.n_nodes()
print "\npendulum_nodes_dmtimes = ", dmtimes.n_nodes()
print "\npendulum_nodes_dgtimes = ", dgtimes.n_nodes()
print "\npendulum_nodes_dfdtimes_aba = ", dfdtimes_aba.n_nodes()
print "\npendulum_nodes_dfdtimes_crba =", dfdtimes_crba.n_nodes()


def dfd_crba_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dfd_crba(q_none, qdot_none, tau_none)
   
def dfdtimes_crba_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        taudot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dfdtimes_crba(q_none, qdot_none, tau_none, qddot_none, taudot_none)

def dfd_aba_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        tau_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dfd_aba(q_none, qdot_none, tau_none)
   
def dfdtimes_aba_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        taudot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dfdtimes_aba(q_none, qdot_none, tau_none, qddot_none, taudot_none)
   
   
   
def dm_func():  
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2  
    dm(q_none)
   
def dg_func():  
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dg(q_none)
   
   
def dmtimes_func():  
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2  
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dmtimes(q_none, qdot_none)
   
def dgtimes_func():  
    for j in range(njoints):
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2        
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    dgtimes(q_none, qdot_none)


def dc_func():  
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    dc(q_none, qdot_none)
   
   
def dctimes_func():  
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2

    dctimes(q_none, qdot_none, qddot_none)
   
   
def did_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    did(q_none, qdot_none, qddot_none)
   
def didtimes_func():
    for j in range(njoints):
        q_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
        qdddot_none[j] = (qmax[j] - qmin[j])*np.random.rand()-(qmax[j] - qmin[j])/2
    didtimes(q_none, qdot_none, qddot_none, qdddot_none)

qdddot_none = [None]*njoints
taudot_none = [None]*njoints





ur5_nodes_did =  234

ur5_nodes_dc =  239

ur5_nodes_dm =  44

ur5_nodes_dg =  29

ur5_nodes_dfd_aba =  390

ur5_nodes_dfd_crba = 378

ur5_nodes_didtimes =  219

ur5_nodes_dctimes =  165

ur5_nodes_dmtimes =  48

ur5_nodes_dgtimes =  32

ur5_nodes_dfdtimes_aba =  290

ur5_nodes_dfdtimes_crba = 292


In [ ]:

ur5_nodes_did =  234

ur5_nodes_dc =  239

ur5_nodes_dm =  44

ur5_nodes_dg =  29

ur5_nodes_dfd_aba =  390

ur5_nodes_dfd_crba = 378



ur5_nodes_didtimes =  219

ur5_nodes_dctimes =  165

ur5_nodes_dmtimes =  48

ur5_nodes_dgtimes =  32

ur5_nodes_dfdtimes_aba =  290

ur5_nodes_dfdtimes_crba = 292


In [8]:
timeit_did = repeat("did_func()", setup = "from __main__ import did_func", repeat = nrepeat, number = nitr)
mediantime_did = median(timeit_did)/nitr*1000000

timeit_dg = repeat("dg_func()", setup = "from __main__ import dg_func", repeat = nrepeat, number = nitr)
mediantime_dg = median(timeit_dg)/nitr*1000000

timeit_dc = repeat("dc_func()", setup = "from __main__ import dc_func", repeat = nrepeat, number = nitr)
mediantime_dc = median(timeit_dc)/nitr*1000000

timeit_dm = repeat("dm_func()", setup = "from __main__ import dm_func", repeat = nrepeat, number = nitr)
mediantime_dm = median(timeit_dm)/nitr*1000000

timeit_dfd_aba = repeat("dfd_aba_func()", setup = "from __main__ import dfd_aba_func", repeat = nrepeat, number = nitr)
mediantime_dfd_aba = median(timeit_dfd_aba)/nitr*1000000

timeit_dfd_crba = repeat("dfd_crba_func()", setup = "from __main__ import dfd_crba_func", repeat = nrepeat, number = nitr)
mediantime_dfd_crba = median(timeit_dfd_crba)/nitr*1000000

   
print "\npendulum_did =", mediantime_did
print "\npendulum_dg =", mediantime_dg
print "\npendulum_dc =", mediantime_dc
print "\npendulum_dm =", mediantime_dm
print "\npendulum_dfd_aba =", mediantime_dfd_aba
print "\npendulum_dfd_crba =", mediantime_dfd_crba


pendulum_did = 28.9194583893

pendulum_dg = 12.6346349716

pendulum_dc = 21.7266082764

pendulum_dm = 12.6569271088

pendulum_dfd_aba = 30.308008194

pendulum_dfd_crba = 30.016541481


pendulum_did = 28.9194583893

pendulum_dg = 12.6240253448

pendulum_dc = 20.5725431442

pendulum_dm = 12.547492981

pendulum_dfd_aba = 29.1010141373

pendulum_dfd_crba = 29.0679931641

In [9]:
timeit_didtimes = repeat("didtimes_func()", setup = "from __main__ import didtimes_func", repeat = nrepeat, number = nitr)
mediantime_didtimes = median(timeit_didtimes)/nitr*1000000

timeit_dgtimes = repeat("dgtimes_func()", setup = "from __main__ import dgtimes_func", repeat = nrepeat, number = nitr)
mediantime_dgtimes = median(timeit_dgtimes)/nitr*1000000

timeit_dctimes = repeat("dctimes_func()", setup = "from __main__ import dctimes_func", repeat = nrepeat, number = nitr)
mediantime_dctimes = median(timeit_dctimes)/nitr*1000000

timeit_dmtimes = repeat("dmtimes_func()", setup = "from __main__ import dmtimes_func", repeat = nrepeat, number = nitr)
mediantime_dmtimes = median(timeit_dmtimes)/nitr*1000000

timeit_dfdtimes_aba = repeat("dfdtimes_aba_func()", setup = "from __main__ import dfdtimes_aba_func", repeat = nrepeat, number = nitr)
mediantime_dfdtimes_aba = median(timeit_dfdtimes_aba)/nitr*1000000

timeit_dfdtimes_crba = repeat("dfdtimes_crba_func()", setup = "from __main__ import dfdtimes_crba_func", repeat = nrepeat, number = nitr)
mediantime_dfdtimes_crba = median(timeit_dfdtimes_crba)/nitr*1000000

print "\npendulum_didtimes =", mediantime_didtimes
print "\npendulum_dgtimes =", mediantime_dgtimes
print "\npendulum_dctimes =", mediantime_dctimes
print "\npendulum_dmtimes =", mediantime_dmtimes
print "\npendulum_dfdtimes_aba =", mediantime_dfdtimes_aba
print "\npendulum_dfdtimes_crba =", mediantime_dfdtimes_crba


pendulum_didtimes = 36.6580486298

pendulum_dgtimes = 20.495057106

pendulum_dctimes = 28.74147892

pendulum_dmtimes = 20.5020904541

pendulum_dfdtimes_aba = 42.9799556732

pendulum_dfdtimes_crba = 44.7340011597



pendulum_didtimes = 36.5949869156

pendulum_dgtimes = 20.383477211

pendulum_dctimes = 28.74147892

pendulum_dmtimes = 20.5020904541

pendulum_dfdtimes_aba = 42.9799556732

pendulum_dfdtimes_crba = 44.7775125504


# Number of Operations

In [15]:
print "\npendulum_nodes_id=", id_u2c.n_nodes()
print "\npendulum_nodes_c=", c_u2c.n_nodes()
print "\npendulum_nodes_m=", m_u2c.n_nodes()
print "\npendulum_nodes_g=", g_u2c.n_nodes()
print "\npendulum_nodes_fdaba", fd_aba_u2c.n_nodes()
print "\npendulum_nodes_fdcrba=", fd_crba_u2c.n_nodes() 

print "\n\npendulum_nodes_id_dq=", id_dq.n_nodes()
print "\npendulum_nodes_c_dq=", c_dq.n_nodes()
print "\npendulum_nodes_m_dq =", m_dq.n_nodes()
print "\npendulum_nodes_g_dq =", g_dq.n_nodes()
print "\npendulum_nodes_fdaba_dq =", fd_aba_dq.n_nodes()
print "\npendulum_nodes_fdcrba_dq =", fd_crba_dq.n_nodes()     

print "\n\npendulum_nodes_id_dqdot=", id_dqdot.n_nodes()
print "\npendulum_nodes_c_dqdot =", c_dqdot.n_nodes()
print "\npendulum_nodes_fdaba_dqdot =", fd_aba_dqdot.n_nodes()
print "\npendulum_nodes_fdcrba_dqdot =", fd_crba_dqdot.n_nodes()     

print "\n\npendulum_nodes_id_dqddot=", id_dqddot.n_nodes()
print "\npendulum_nodes_fdaba_dtau =", fd_aba_dtau.n_nodes()
print "\npendulum_nodes_fdcrba_dtau =", fd_crba_dtau.n_nodes()   


pendulum_nodes_id= 110

pendulum_nodes_c= 79

pendulum_nodes_m= 34

pendulum_nodes_g= 18

pendulum_nodes_fdaba 142

pendulum_nodes_fdcrba= 132


pendulum_nodes_id_dq= 166

pendulum_nodes_c_dq= 130

pendulum_nodes_m_dq = 44

pendulum_nodes_g_dq = 29

pendulum_nodes_fdaba_dq = 257

pendulum_nodes_fdcrba_dq = 259


pendulum_nodes_id_dqdot= 149

pendulum_nodes_c_dqdot = 130

pendulum_nodes_fdaba_dqdot = 199

pendulum_nodes_fdcrba_dqdot = 191


pendulum_nodes_id_dqddot= 37

pendulum_nodes_fdaba_dtau = 55

pendulum_nodes_fdcrba_dtau = 43



pendulum_nodes_id= 110

pendulum_nodes_c= 79

pendulum_nodes_m= 34

pendulum_nodes_g= 18

pendulum_nodes_fdaba 142

pendulum_nodes_fdcrba= 132


pendulum_nodes_id_dq= 166

pendulum_nodes_c_dq= 130

pendulum_nodes_m_dq = 44

pendulum_nodes_g_dq = 29

pendulum_nodes_fdaba_dq = 257

pendulum_nodes_fdcrba_dq = 259


pendulum_nodes_id_dqdot= 149

pendulum_nodes_c_dqdot = 130

pendulum_nodes_fdaba_dqdot = 199

pendulum_nodes_fdcrba_dqdot = 191


pendulum_nodes_id_dqddot= 37

pendulum_nodes_fdaba_dtau = 55

pendulum_nodes_fdcrba_dtau = 43

